tests/mgbench/client.cpp

```diff
@@ -220,7 +224,106 @@ nlohmann::json LatencyStatistics(std::vector<std::vector<double>> &worker_query_
   return statistics;
 }
 
-void Execute(
+void ExecuteTimeDependentWorkload(
+    const std::vector<std::pair<std::string, std::map<std::string, memgraph::communication::bolt::Value>>> &queries,
+    std::ostream *stream) {
+  std::vector<std::thread> threads;
+  threads.reserve(FLAGS_num_workers);
+
+  std::vector<uint64_t> worker_retries(FLAGS_num_workers, 0);
+  std::vector<Metadata> worker_metadata(FLAGS_num_workers, Metadata());
+  std::vector<double> worker_duration(FLAGS_num_workers, 0.0);
+  std::vector<std::vector<double>> worker_query_durations(FLAGS_num_workers);
+
+  // Start workers and execute queries.
+  auto size = queries.size();
+  std::atomic<bool> run(false);
+  std::atomic<uint64_t> ready(0);
+  std::atomic<uint64_t> position(0);
+
+  std::chrono::duration<double> time_limit = std::chrono::seconds(FLAGS_time_dependent_execution);
+  for (int worker = 0; worker < FLAGS_num_workers; ++worker) {
+    threads.push_back(std::thread([&, worker]() {
+      memgraph::io::network::Endpoint endpoint(FLAGS_address, FLAGS_port);
+      memgraph::communication::ClientContext context(FLAGS_use_ssl);
+      memgraph::communication::bolt::Client client(context);
+      client.Connect(endpoint, FLAGS_username, FLAGS_password);
+
+      ready.fetch_add(1, std::memory_order_acq_rel);
+      while (!run.load(std::memory_order_acq_rel))
+        ;
+
+      auto &retries = worker_retries[worker];
+      auto &metadata = worker_metadata[worker];
+      auto &duration = worker_duration[worker];
+      auto &query_duration = worker_query_durations[worker];
+
+      memgraph::utils::Timer worker_timer;
+      while (true) {
+        double total_workload_duration =
+            std::accumulate(worker_duration.begin(), worker_duration.end(), 0.0) / FLAGS_num_workers;
```

Comment:
Why accumulating her, isn't the `total_time` 1-1 mapped to the `worker_timer`? + Iterating over an array in the middle of a hot loop seems expansive. Where is `worker_duration` updated?

In general, this code seems more complex than it should be.

Reply:
Rewrite this to avoid accumulating.